test/e2e/storage/subpath.go

```diff
@@ -168,203 +137,372 @@ var _ = utils.SIGDescribe("Subpath", func() {
-				setInitCommand(pod, fmt.Sprintf("touch %s", subPathDir))
+func (s *subPathTestCase) createTestInput(
+	testPattern testPattern,
+	testResource testResource,
+) testInput {
+	if r, ok := testResource.(*subPathTestResource); ok {
+		driver := r.driver
+		driverInfo := driver.getDriverInfo()
+		f := driverInfo.f
+		subPath := f.Namespace.Name
+		subPathDir := filepath.Join(volumePath, subPath)
 
-				// Write the file in the volume from container 1
-				setWriteCommand(subPathDir, &pod.Spec.Containers[1])
+		return subPathTestInput{
+			f:                 f,
+			subPathDir:        subPathDir,
+			filePathInSubpath: filepath.Join(volumePath, fileName),
+			filePathInVolume:  filepath.Join(subPathDir, fileName),
+			volType:           r.volType,
+			pod:               r.pod,
+			volSource:         r.volSource,
+			roVol:             r.roVolSource,
+		}
+	}
 
-				// Read it from inside the subPath from container 0
-				pod.Spec.Containers[0].VolumeMounts[0].ReadOnly = true
-				testReadFile(f, volumePath, pod, 0)
-			})
+	framework.Failf("Fail to convert testResource to subPathTestResource")
+	return nil
+}
 
-			It("should support existing directories when readOnly specified in the volumeSource", func() {
-				roVol := vol.getReadOnlyVolumeSpec()
-				if roVol == nil {
-					framework.Skipf("Volume type %v doesn't support readOnly source", curVolType)
-				}
+func (s *subPathTestCase) getTestFunc() func(*testInput) {
+	return testSubPathWithTestInput
+}
 
-				// Initialize content in the volume while it's writable
-				initVolumeContent(f, pod, filePathInVolume, filePathInSubpath)
+func testSubPathWithTestInput(testInput *testInput) {
+	Context("-", func() {
+		var (
+			t  subPathTestInput
+			ok bool
+		)
 
-				// Set volume source to read only
-				pod.Spec.Volumes[0].VolumeSource = *roVol
+		BeforeEach(func() {
+			testInputVal := *testInput
+			if t, ok = testInputVal.(subPathTestInput); !ok {
+				framework.Failf("Fail to convert testInput to subPathTestInput")
+			}
+		})
 
-				// Read it from inside the subPath from container 0
-				testReadFile(f, filePathInSubpath, pod, 0)
-			})
+		testSubPath(&t)
+	})
+}
 
-			It("should fail for new directories when readOnly specified in the volumeSource", func() {
-				roVol := vol.getReadOnlyVolumeSpec()
-				if roVol == nil {
-					framework.Skipf("Volume type %v doesn't support readOnly source", curVolType)
+type subPathTestResource struct {
+	driver      testDriver
+	volType     string
+	volSource   *v1.VolumeSource
+	pvc         *v1.PersistentVolumeClaim
+	pv          *v1.PersistentVolume
+	roVolSource *v1.VolumeSource
+	sc          *storagev1.StorageClass
+	pod         *v1.Pod
+}
+
+func (s *subPathTestResource) setup(driver testDriver, testPattern testPattern) {
+	s.driver = driver
+	driverInfo := s.driver.getDriverInfo()
+	f := driverInfo.f
+	cs := f.ClientSet
+	fsType := testPattern.fsType
+
+	skipCreatingResourceForProvider(testPattern)
+
+	switch testPattern.testVolType {
```

Comment:
I see this is already being done [here](https://github.com/kubernetes/kubernetes/pull/66577/files#diff-b2926bcb65edc51156080c46e7e5fcf1R320).  Do we need to repeat it again?  Or can we try to have common wrappers for the shared parts?

Reply:
```skipCreatingResourceForProvider``` is removed from ```setup```.
Necessary skip will be done ```skipUnsupportedTest``` in test loop, by using ```testDriver```s' and ```testSuite```s' ```skipUnsupportedTest```.